Package rewrite - convert Client to an Interface

marz619 opened the pull request:
Changes the exposed `Client` and returns it as an interface rather than a struct type. Overall, this is an attempt at cleaning up the interface and only exposing the barebones in an attempt to maintain a level of consistency and _future-proofing_ for projects that are using this package.

## Changes

Each subsequent commit is fairly well encapsulated and can be looked at individually to see the overall changes being made.

* update the License
* update the README
* public `Client` is now an `interface` that will implement the necessary  HTTP verb methods for use by the caller. This replicates the current working API that was previously exported as a `struct{}` type.  This interface is implemented internally with a `type client struct` that holds all the configuration options and a reference to the underlying `*http.Client`.

```go
type Client interface {
	// Client returns the underlying *http.Client
	Client() *http.Client

	// Close all Idle connections
	Close()

	// HTTP Request Methods //

	// Do is the generic HTTP request method. The two string parameters in
	// order are the HTTP Method, and the URL to request respectively
	Do(ctx context.Context, rh ResponseHandler, method, url string, body io.Reader, opts ...RequestOption) error

	// GET method
	Get(ctx context.Context, rh ResponseHandler, url string, opts ...RequestOption) error

	// POST method
	Post(ctx context.Context, rh ResponseHandler, url string, body io.Reader, opts ...RequestOption) error
}
```

* introduce the `Option` type and move all Client configuration option functions to their own file `client_option.go`
* consolidate all the request methods into the same file `request.go`
* running golint, go vet and updating/adding tests where necessary
* Adding a `go.mod` file
* An attempt (seemingly successful) to fix #1 - 12db943

marz619 committed:
update README, LICENSE, and .gitignore

marz619 committed:
REDUX - rewrite Client as an Interface

convert the Client struct into an interface
* implemented internally using the client struct

adds an Option type used to configure the underlying *http.Client
*
https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
* DisableHTTP2 option is configured via Transport.ForceAttemptHTTP2

move all _request_ methods Do, Get, Post & tests to requests.go &
requests_test.go

marz619 committed:
add go.mod

marz619 committed:
sort client config Options alphabetically

marz619 committed:
add RoundTripperFunc convenience type

It is similar to http.HandlerFunc in that it implements the RoundTrip
method, thereby satisifying the RoundTripper interface

marz619 committed:
update client Option comments

marz619 committed:
attempt to fix #1 - updating SetHeaders, and tests

marz619 commented on client_option.go:
```diff
@@ -0,0 +1,183 @@
+	}
+}
+
+// RoundTripperFunc is like http.HandlerFunc, but for RoundTripper interface.
```
```suggestion
// RoundTripperFunc is like http.HandlerFunc for http.RoundTripper interface.
```

rafikdraoui commented on client.go:
```diff
@@ -13,244 +13,123 @@ import (
 )
 
-//ErrInvalidValue signifies that an invaild value was given to configartion option
+// ErrInvalidValue signifies that an invaild value was given to configartion option
```
```suggestion
// ErrInvalidValue signifies that an invalid value was given to configuration option
```

rafikdraoui commented on client.go:
```diff
@@ -13,244 +13,123 @@ import (
 
-func (c *Client) setDefaults() {
-	c.headers = make(http.Header)
+// set *sensible* default values on the *client
```
```suggestion
// set sensible default values on the *client
```

The `*` can be a bit confusing when it's also used to denote a pointer type.

rafikdraoui commented on request_option_test.go:
```diff
@@ -0,0 +1,74 @@
+			// add X-Test if we're not already checking for it
+			if _, ok := tc.exp["X-Test"]; !ok {
+				tc.exp["X-Test"] = headers.Values("X-Test")
+			}
```
When I comment out this section, the tests still pass. So either it isn't needed, or else we are not truly exercising some cases.

marz619 replied:
made the Test more explicit by adding multiple maps to denote exactly what we're adding (extra), and what's expected here: b899ea3

rafikdraoui commented on request_option_test.go:
```diff
@@ -0,0 +1,74 @@
+		// make the request
+		err = c.Get(context.Background(), NoopResponseHandler, s.URL, SetHeaders(extra))
+		if err != nil {
+			t.Errorf("trouble with set headers roundtripper unexpected -> %v", err)
```
```suggestion
			t.Errorf("unexpected error with SetHeaders: %v", err)
```

rafikdraoui commented on request_option.go:
```diff
@@ -30,15 +31,39 @@ func DelHeaders(headers http.Header) RequestOption {
-			var val []string
-			copy(val, v)
-			req.Header[k] = val
+		for k, vs := range headers {
+			req.Header.Del(k)
+			for i := range vs {
+				req.Header.Add(k, vs[i])
+			}
```
👍 I think that should do the trick: reading the docs, it seems that accessing the Header map directly bypasses canonicalization of key names, so using the `Add`/`Set`/`Del`/etc. methods should take care of the issue.

I'm assuming that this was the issue: the original bug report is a bit vague ("SetHeader is broken") 😅

rafikdraoui reviewed (commented):
The code looks good to me. I think the next step is to update some projects to use this branch as a pre-release version, and see how things go!

marz619 committed:
rename ErrInvalidValue to ErrInvalidOptionValue

marz619 committed:
PR suggestion/fixes

- fix comments
- make the testCase for TestSetHeaders more explicit
  - use t.Run to run cases as sub-tests
  - add an extra map that contains the values to use with SetHeaders
  - add an expected map to denote the expected Header values to be
    receieved by the test Server

rafikdraoui commented on request_option_test.go:
```diff
@@ -0,0 +1,100 @@
+				t.Errorf("unexpected error with SetHeaders: %v", err)
+			}
+			// close client & server
+			c.Close()
+			s.Close()
```
These could be deferred when `c` and `s` are defined instead. 

I tried making a GitHub suggestion for the change, but it couldn't compute a nice clean diff for it, so here's a patch generated from the git CLI instead!

```diff
diff --git a/request_option_test.go b/request_option_test.go
index ff56f83..9bb24ec 100644
--- a/request_option_test.go
+++ b/request_option_test.go
@@ -66,6 +66,7 @@ func TestSetHeaders(t *testing.T) {
 			if err != nil {
 				t.Errorf("trouble when creating the client: %v", err)
 			}
+			defer c.Close()
 
 			// startup a server with handler to check that received request headers
 			// are correct
@@ -79,6 +80,7 @@ func TestSetHeaders(t *testing.T) {
 					}
 				}
 			}))
+			defer s.Close()
 
 			// create our extra headers
 			extra := make(http.Header)
@@ -92,9 +94,6 @@ func TestSetHeaders(t *testing.T) {
 			if err != nil {
 				t.Errorf("unexpected error with SetHeaders: %v", err)
 			}
-			// close client & server
-			c.Close()
-			s.Close()
 		})
 	}
 }
```

marz619 committed:
use defer to close client/server in TestSetHeaders

Co-authored-by: Rafik Draoui <rdraoui@gadventures.com>
